q2_sample_classifier/utilities.py

```diff
@@ -346,6 +346,15 @@ def _fit_estimator(features, targets, estimator, n_estimators=100, step=0.05,
     if optimize_feature_selection:
         estimator.rfe_scores = rfe_scores
 
+    # methods cannot output an empty importances artifact; only KNN has no
+    # feature importance, but just warn and output all features as
+    # importance = 1
```

Comment:
Can you add a `TODO` note here to drop this whenever we get around to supporting optional outputs? I don't see a framework issue for this to xref with, and I am hesitant to make one ATM.

Reply:
As well - is there a way to promote the warning into one of the visualizations produced, too?

Reply:
the visualizations produced no longer contain lists of important features, so there is not really a relevant place to post this warning. I suppose I could toss it into the `summary` viz, but this would not have the intended effect of keeping users from trying to do anything meaningful with the `importances` (which most users probably won't anyway). Thoughts?

Reply:
Ah good point. Bummer, I guess one option is a comment line in the the importances file format itself, but you wouldn't see it when running `metadata tabulate`, which defeats the purpose (maybe I am wrong on that though, worth testing).

Reply:
Just tested — comments are not displayed.

I could make the importances `None` or `Null` or something along those lines, which will just make that output useless — but I don't know what consequences that could have, e.g., if this is used in a plugin.

Reply:
The `None` output is an interesting idea --- @ebolyen, what do you think?

> but I don't know what consequences that could have

This sounds like the "empty feature table" situation - people get less-than-helpful errors when feeding empty tables into downstream methods.

Reply:
Is this a good example of a NaN artifact? I think that's our most realistic way of implementing optional outputs, which seems like the core problem here.

For reference, a NaN artifact could be used anywhere, but would cause any actions to immediately return NaN artifact outputs as well.

Reply:
What is a NaN artifact? Would the values be NaN, or would it be an empty file?

Reply:
Probably empty file. It's more an artifact that works in QIIME 2 the same way NaN works with floating point operations.

Reply:
Or no file whatsoever

Reply:
Also to clarify, this doesn't exist yet. I'm just building up reasons it maybe should or shouldn't exist.

Reply:
don't use this as a reason then — only knn estimators have this issue and users aren't likely to use them.

Reply:
In the meantime, some kind of placeholder value sounds good to me. I think I like None better than 1 - it feels less likely to lead someone astray (I could see someone seeing 1s and thinking "jackpot!" - that someone is me).